### PR TITLE
[Return-types Fix] Update `batch_params` and `batch_input` transforms

### PR DESCRIPTION
**Pull request by eddddddy**
**Context:**
Part of the ongoing effort to update the `batch_transforms` for the new return types system.

**Description of the Change:**
Update the `batch_params` and `batch_input` transforms to stack the results at the leaf nodes of the nested tuple structure. More work needs to be done to integrate parameter broadcasting.

### CODE REVIEW

File: tests/returntypes/transforms/test_batch_input_new.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        return qml.probs(wires=[0, 1, 2])
+
+    indiv_res = []
+    for state in data:
+        indiv_res.append(circuit2(state, weights))
+    assert np.allclose(res, indiv_res)
+
+
+def test_multi_returns():
+    """Test that batching works for a simple circuit with multiple returns"""
```
Comment (eddddddy) [inline tests/returntypes/transforms/test_batch_input_new.py:181]: Added this test and the next two to test for multiple returns, shot vectors, and combined.

File: tests/returntypes/transforms/test_batch_params_new.py [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+        return qml.probs(wires=[0, 1, 2, 3])
+
+    indiv_res = []
+    for state, weight in zip(data, weights):
+        indiv_res.append(circuit2(state, weight))
+    assert np.allclose(res, indiv_res)
+
+
+def test_multi_returns():
+    """Test that batching works for a simple circuit with multiple returns"""
```
Comment (eddddddy) [inline tests/returntypes/transforms/test_batch_params_new.py:184]: Same with these tests.

File: tests/returntypes/transforms/test_batch_input_new.py (hunk 1 above)
Comment (rmoyard) [reply tests/returntypes/transforms/test_batch_input_new.py:181]: Thanks for pointing this out!

File: tests/returntypes/transforms/test_batch_params_new.py (hunk 2 above)
Comment (rmoyard) [reply tests/returntypes/transforms/test_batch_params_new.py:184]: Thanks for pointing this out!

**Review by rmoyard — COMMENTED** [review]
It looks great @eddddddy 💯 I suggest to try adding some test for jacobian and multiple measurements if possible depending on the framework. Happy to approve after trying that!

File: tests/returntypes/transforms/test_batch_params_new.py [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+    x = np.linspace(0.1, 0.5, batch_size, requires_grad=True)
+
+    res = qml.grad(cost)(x)
+    expected = -np.sin(0.1) * np.sin(x)
+    assert np.allclose(res, expected, atol=tol, rtol=0)
+
+
+@pytest.mark.jax
+@pytest.mark.parametrize("diff_method", ["backprop", "adjoint", "parameter-shift"])
+def test_jax(diff_method, tol):
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply tests/returntypes/transforms/test_batch_params_new.py]: Added tests with multiple measurements for all interfaces.

File: tests/returntypes/transforms/test_batch_input_new.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    assert len(res) == 5
+    assert all(isinstance(shot_res, tuple) for shot_res in res)
+    assert all(len(shot_res) == 2 for shot_res in res)
+    assert all(shot_res[0].shape == (batch_size,) for shot_res in res)
+    assert all(shot_res[1].shape == (batch_size, 4) for shot_res in res)
+
+
+@pytest.mark.autograd
+@pytest.mark.parametrize("diff_method", ["backprop", "adjoint", "parameter-shift"])
+def test_autograd(diff_method, tol):
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply tests/returntypes/transforms/test_batch_input_new.py]: Yes, it looks like this transform was not tested for all interfaces. I added some tests also for multiple returns.

**Review by rmoyard — APPROVED** [review]
Thanks for the test addition, it looks good to me!

File: pennylane/interfaces/tensorflow.py [hunk 5]
```diff
@@ -111,19 +111,20 @@ def _jac_restructured(jacs, tapes):
...
     jacs_nested = []
-    for i, tape in enumerate(tapes):
+    for tape in tapes:
         num_meas = len(tape.measurements)
         num_params = len(tape.trainable_params)
 
         tape_jacs = tuple(jacs[start : start + num_meas * num_params])
         tape_jacs = tuple(
-            tuple(tape_jacs[i * num_params : (i + 1) * num_params]) for i in range(num_meas)
+            tuple(tape_jacs[j * num_params : (j + 1) * num_params]) for j in range(num_meas)
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/interfaces/tensorflow.py:120]: Does this fix belong in this PR? :thinking: What was the original reason for the change here?

File: pennylane/interfaces/tensorflow.py [hunk 6]
```diff
@@ -111,19 +111,20 @@ def _jac_restructured(jacs, tapes):
...
         )
 
         while isinstance(tape_jacs, tuple) and len(tape_jacs) == 1:
             tape_jacs = tape_jacs[0]
 
         jacs_nested.append(tape_jacs)
+        start += num_meas * num_params
```
Comment (antalszava) [inline pennylane/interfaces/tensorflow.py:127]: Was `start` not changed previously? :thinking:

File: pennylane/transforms/batch_params.py [hunk 7]
```diff
@@ -21,6 +21,22 @@
 from .batch_transform import batch_transform
 
 
+def _nested_stack(res):
+    """
+    Given a list of identical nested tuple structures, stack the arrays at the leaves
+    """
+    # for some reason pylint thinks qml.numpy.builtins is a dict
+    # pylint: disable=no-member
```
Comment (antalszava) [inline pennylane/transforms/batch_params.py:29]: Maybe the `.pylintrc` file could be updated to account for this.

File: tests/returntypes/transforms/test_batch_input_new.py [hunk 8]
```diff
@@ -0,0 +1,670 @@
...
+        with tf.GradientTape() as tape:
+            loss = tf.reduce_sum(circuit(input, x))
+
+        res = tape.gradient(loss, x)
+        expected = -np.sin(0.1) * tf.reduce_sum(tf.sin(input))
+        print(circuit(input, x))
+        print(np.cos(0.1) * tf.sin(input))
+        print(loss)
+        print(res)
+        print(expected)
```
Comment (antalszava) [inline tests/returntypes/transforms/test_batch_input_new.py:400]:
```suggestion
```

File: tests/returntypes/transforms/test_batch_input_new.py [hunk 9]
```diff
@@ -0,0 +1,670 @@
...
+                        np.cos((input + x) / 2) ** 2,
+                        np.zeros_like(input),
+                        np.zeros_like(input),
+                        np.sin((input + x) / 2) ** 2,
+                    ]
+                )
+            ),
+        )
+
+        assert len(res) == 2
```
Comment (antalszava) [inline tests/returntypes/transforms/test_batch_input_new.py:493]: Let's check the type of `res` too.

File: tests/returntypes/transforms/test_batch_input_new.py [hunk 10]
```diff
@@ -0,0 +1,670 @@
...
+                        -jnp.sin(input + x) / 2,
+                        jnp.zeros_like(input),
+                        jnp.zeros_like(input),
+                        jnp.sin(input + x) / 2,
+                    ]
+                )
+            ),
+        )
+
+        assert len(grad) == 2
```
Comment (antalszava) [inline tests/returntypes/transforms/test_batch_input_new.py:512]: Let's check the type of `grad` too.

File: tests/returntypes/transforms/test_batch_input_new.py [hunk 11]
```diff
@@ -0,0 +1,670 @@
...
+                    np.zeros_like(input),
+                    np.sin(input + x) / 2,
+                ]
+            )
+        )
+        assert qml.math.allclose(grad, expected, atol=tol, rtol=0)
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("diff_method", ["backprop", "parameter-shift"])
+    def test_jax(self, diff_method, tol):
```
Comment (antalszava) [inline tests/returntypes/transforms/test_batch_input_new.py:459]: How about checking jitting too? (or adding a TODO for it here)

**Review by antalszava — COMMENTED** [review]
Just a couple of questions, overall looking good! :100:

File: pennylane/interfaces/tensorflow.py (hunk 6 above)
Comment (eddddddy) [reply pennylane/interfaces/tensorflow.py:127]: It was not. Previously the test file only calls `qml.execute` on single tapes, so this was a bug that slipped through. Batching requires calling `qml.execute` with multiple tapes now.

File: pennylane/interfaces/tensorflow.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/interfaces/tensorflow.py:120]: Not sure what I was thinking before but there was already a variable named `i`. Actually, with this change we removed the `enumerate(tapes)` and just iterate over `tapes` itself, so I can change it back to `i`.

File: tests/returntypes/transforms/test_batch_input_new.py (hunk 11 above)
Comment (antalszava) [reply tests/returntypes/transforms/test_batch_input_new.py:493]: https://github.com/PennyLaneAI/pennylane/pull/3445 should be fine now :+1:

**Pull request merged**